### DISCUSSION TITLE
Need to continue streaming - even in background. Is it possible?

### DISCUSSION THREAD

**Discussion by dhruvdroid** [ThibaultBee/StreamPack#150] [Q&A]
Hi @ThibaultBee, 
 
Currently, I am working to integrate StreamPack for the live-streaming app. 
While checking the code base I got to know we have the `StreamerLifeCycleObserver` to which we provide the `lifecycle.addObserver(streamerLifeCycleObserver)`

```
open class StreamerLifeCycleObserver(var streamer: IStreamer) : DefaultLifecycleObserver {
    override fun onPause(owner: LifecycleOwner) {
        streamer.getCameraStreamer()?.stopPreview()
        runBlocking {
            streamer.stopStream()
        }
        streamer.getLiveStreamer()?.let {
            if (it.isConnected) {
                it.disconnect()
            }
        }
    }

    override fun onDestroy(owner: LifecycleOwner) {
        streamer.release()
    }
}
```

The assumption here was, in case I don't provide this `streamerLifeCycleObserver` then devs needs to manually handle the stop and release the stream in `onPause` and `onDestroy`

Even after commenting on the code i.e.`lifecycle.addObserver(streamerLifeCycleObserver)` the moment onPause gets trigger following logs are printed - 
```
MediaCodecEncoder        D  Not running
MediaCodecEncoder        D  Not running
AudioRecord              D  stop(4293): mActive:1
AudioRecord              D  mAudioRecord->stop()
AudioRecord              D  AudioRecordThread pause()
AudioRecord              D  stop() end
OplusCCodec              D  initiateShutdown [386]: (0xb40000767df4d480) keepComponentAllocated=0
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer
CCodecBufferChannel      D  [c2.android.aac.encoder#51] MediaCodec discarded an unknown buffer 	
```

The requirement here is - 
1. The live stream should continue in the background until we are stopping it via End Live button explicitly. 
2. If the above logs can be the root cause to stop the stream if we open the dialog or screen on top of the `io.github.thibaultbee.streampack.views.PreviewView`. How can I continue streaming in that case? Is there a document I can refer or example.

**Comment by ThibaultBee** [comment]
If you want to implement another lifecycle, you don't have to use `StreamerLifeCycleObserver`.

If you want to live stream in background, you should use an Android service. As an example, you can have a look on screen recorder services provide by StreamPack. I have not tested it because it does not seem supported by old devices.

**Reply by dhruvdroid** [reply]
Thanks @ThibaultBee for the quick response. 

While further debugging if I remove the lifecycle observer i.e. `StreamerLifeCycleObserver`. At least `streamer.preview` and `stopStream` is not happening.  

As per the `PreviewClass` code, we are calling the `streamer?.stopPreview()` on `surfaceDestroyed` callback. 

```
private inner class StreamerHolderCallback : SurfaceHolder.Callback {
        override fun surfaceCreated(holder: SurfaceHolder) {
            startPreviewIfReady()
        }

        override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) =
            Unit

        override fun surfaceDestroyed(holder: SurfaceHolder) {
            streamer?.stopPreview()
        }
    }
```

1. Is there a way I can avoid calling `streamer?.stopPreview()` and resume/continue the stream from here?
2. I tried inheriting the `PreviewClass` as it is a final class and cannot be subclassed.

FYI - don't want to run the streaming in the background forever, just opening any activity or dialog over `PreviewView` is automatically stops the stream. Just want to avoid the same. 

Any suggestion on the above @ThibaultBee?

**Reply by dhruvdroid** [reply]
To overcome the `stopPreview` being called and ended stream - I am planning to do the following - 
```
private fun continueStreamingPostPinningProduct() {
        if(streamer.isConnected) {
            launchCoroutine {
                streamer.disconnect()
                binding.previewView.streamer = streamer
                startStream(streamingUrl)
            }
        }
    }
```

Do let me know if I am missing any important steps to continue streaming - this seems to be working fine for now.

**Reply by ThibaultBee** [reply]
I guess the best solution is to dissociate preview and streaming. A `stopPreview` should not cause a `stopStream`.
This requires internal changes that I can't tackle right now.

I don't fully understand your solution because there are missing parts. Anyway, you don't even have to call `disconnect`, you can directly call `streamer.startStream()` (the `startStream` without parameters)

**Reply by cristiAndreiTarasi** [reply]
Hi. To the guy that asked, probaly disabling stopStream an disconect methods in StreamerLifeCycleObserver would do the trick. Seems naive but I think it would maintain the connection in background. 

```
open class StreamerLifeCycleObserver(var streamer: IStreamer) : DefaultLifecycleObserver {
    override fun onPause(owner: LifecycleOwner) {
        // Stop preview to save resources, but do not stop the stream.
        streamer.getCameraStreamer()?.stopPreview()

        // Comment this code out to prevent disrupting the connection
        /*runBlocking {
            streamer.stopStream()
        }
        streamer.getLiveStreamer()?.let {
            if (it.isConnected) {
                it.disconnect()
            }
        }*/
    }

    override fun onDestroy(owner: LifecycleOwner) {
        streamer.release()
    }

    override fun onResume(owner: LifecycleOwner) {
        // Resume the preview when the app comes back to the foreground
        streamer.getCameraStreamer()?.startPreview()
    }
}
```
